File: 2025-2/sesion-de-clase-11/SoftProgPersistencia/DAOImpl/Ventas/LineaOrdenVentaDAOImpl.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using PUCP.SoftProg.Modelo.Ventas;
using PUCP.SoftProg.Persistencia.DAO.Ventas;
using PUCP.SoftProg.Persistencia.DAOImpl.Almacen;

namespace PUCP.SoftProg.Persistencia.DAOImpl.Ventas {
    public class LineaOrdenVentaDAOImpl : TransactionalBaseDAOImpl<LineaOrdenVenta>, ILineaOrdenVentaDAO {
        protected override DbCommand ComandoCrear(DbConnection conn, LineaOrdenVenta lineaOrdenVenta) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarLineaOrdenVenta";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idOrdenVenta", DbType.Int32, lineaOrdenVenta.OrdenVenta.Id);
            this.AgregarParametroEntrada(cmd, "@p_idProducto", DbType.Int32, lineaOrdenVenta.Producto.Id);
            this.AgregarParametroEntrada(cmd, "@p_cantidad", DbType.Int32, lineaOrdenVenta.Cantidad);
            this.AgregarParametroEntrada(cmd, "@p_subTotal", DbType.Double, lineaOrdenVenta.SubTotal);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, lineaOrdenVenta.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand ComandoActualizar(DbConnection conn, LineaOrdenVenta lineaOrdenVenta) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarLineaOrdenVenta";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idOrdenVenta", DbType.Int32, lineaOrdenVenta.OrdenVenta.Id);
            this.AgregarParametroEntrada(cmd, "@p_idProducto", DbType.Int32, lineaOrdenVenta.Producto.Id);
            this.AgregarParametroEntrada(cmd, "@p_cantidad", DbType.Int32, lineaOrdenVenta.Cantidad);
            this.AgregarParametroEntrada(cmd, "@p_subTotal", DbType.Double, lineaOrdenVenta.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, lineaOrdenVenta.Id);

            return cmd;
        }

        protected override DbCommand ComandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarLineaOrdenVenta";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeer(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarLineaOrdenVentaPorId";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeerTodos(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarLineasOrdenVenta";
            cmd.CommandType = CommandType.StoredProcedure;
            return cmd;
        }

        protected override LineaOrdenVenta MapearModelo(DbDataReader reader) {
            return new LineaOrdenVenta {
                Id = Convert.ToInt32(reader["id"]),
                OrdenVenta = new OrdenVentaDAOImpl().Leer(Convert.ToInt32(reader["idOrdenVenta"])),
                Producto = new ProductoDAOImpl().Leer(Convert.ToInt32(reader["idProducto"])), 
                Cantidad = Convert.ToInt32(reader["cantidad"]), 
                SubTotal = Convert.ToDouble(reader["subTotal"]), 
                IsActive = Convert.ToBoolean(reader["activo"])
            };
        }

        protected DbCommand ComandoLeerTodosPorOrden(DbConnection conn, int idOrden, 
            DbTransaction transaccion = null) {

            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarLineasPorOrdenVenta";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idOrdenVenta", DbType.Int32, idOrden);

            if (transaccion != null) {
                cmd.Transaction = transaccion;
            }

            return cmd;
        }

        public List<LineaOrdenVenta> LeerTodosPorOrden(int idOrden, DbTransaction transaccion = null) {
            DbConnection conn = transaccion?.Connection;
            if (conn == null) {
                return EjecutarComando(c => LeerTodosPorOrden(idOrden, c));
            }

            using (DbCommand cmd = this.ComandoLeerTodosPorOrden(conn, idOrden, transaccion))
            using (DbDataReader reader = cmd.ExecuteReader()) {
                List<LineaOrdenVenta> modelos = new List<LineaOrdenVenta>();
                while (reader.Read()) {
                    modelos.Add(this.MapearModelo(reader));
                }
                return modelos;
            }
        }

        private List<LineaOrdenVenta> LeerTodosPorOrden(int idOrden, DbConnection conn) {
            using (DbCommand cmd = this.ComandoLeerTodosPorOrden(conn, idOrden))
            using (DbDataReader reader = cmd.ExecuteReader()) {
                List<LineaOrdenVenta> modelos = new List<LineaOrdenVenta>();
                while (reader.Read()) {
                    modelos.Add(this.MapearModelo(reader));
                }
                return modelos;
            }
        }
    }
}

File: 2025-2/sesion-de-clase-22/java/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/ventas/LineaOrdenVentaDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.ventas;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import java.util.ArrayList;
import java.util.List;
import pe.edu.pucp.inf30.softprog.dao.ventas.LineaOrdenVentaDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.TransaccionalBaseDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.almacen.ProductoDAOImpl;
import pe.edu.pucp.inf30.softprog.modelo.ventas.LineaOrdenVenta;

/**
 *
 * @author eric
 */
public class LineaOrdenVentaDAOImpl extends TransaccionalBaseDAO<LineaOrdenVenta> 
        implements LineaOrdenVentaDAO {

    @Override
    protected PreparedStatement comandoCrear(Connection conn, 
            LineaOrdenVenta modelo) throws SQLException {
        
        String sql = "{call insertarLineaOrdenVenta(?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_idOrdenVenta", modelo.getOrdenVenta().getId());
        cmd.setInt("p_idProducto", modelo.getProducto().getId());
        cmd.setInt("p_cantidad", modelo.getCantidad());
        cmd.setDouble("p_subTotal", modelo.getSubTotal());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, 
            LineaOrdenVenta modelo) throws SQLException {
        
        String sql = "{call modificarLineaOrdenVenta(?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_idOrdenVenta", modelo.getOrdenVenta().getId());
        cmd.setInt("p_idProducto", modelo.getProducto().getId());
        cmd.setInt("p_cantidad", modelo.getCantidad());
        cmd.setDouble("p_subTotal", modelo.getSubTotal());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.setInt("p_id", modelo.getId());
        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, 
            Integer id) throws SQLException {
         
        String sql = "{call eliminarLineaOrdenVenta(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, 
            Integer id) throws SQLException {
        
        String sql = "{call buscarLineaOrdenVentaPorId(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) 
            throws SQLException {
        
        String sql = "{call listarLineasOrdenVenta()}";
        CallableStatement cmd = conn.prepareCall(sql);
        return cmd;
    }

    @Override
    protected LineaOrdenVenta mapearModelo(ResultSet rs) throws SQLException {
        LineaOrdenVenta lineaOrdenVenta = new LineaOrdenVenta();
        lineaOrdenVenta.setId(rs.getInt("id"));
        lineaOrdenVenta.setOrdenVenta(
                new OrdenVentaDAOImpl().leer(rs.getInt("idOrdenVenta")));
        lineaOrdenVenta.setProducto(
                new ProductoDAOImpl().leer(rs.getInt("idProducto")));
        lineaOrdenVenta.setCantidad(rs.getInt("cantidad"));
        lineaOrdenVenta.setSubTotal(rs.getDouble("subTotal"));
        lineaOrdenVenta.setActivo(rs.getBoolean("activo"));
        return lineaOrdenVenta;
    }

    protected PreparedStatement comandoLeerTodosPorOrden(Connection conn, 
            int idOrden) throws SQLException {
        
        String sql = "{call listarLineasPorOrdenVenta(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_idOrdenVenta", idOrden);
        return cmd;
    }
    
    public List<LineaOrdenVenta> leerTodosPorOrden(int idOrden) {
        return ejecutarComando(conn -> leerTodosPorOrden(idOrden, conn));
    }
    
    @Override
    public List<LineaOrdenVenta> leerTodosPorOrden(int idOrden, Connection conn) {
        try (PreparedStatement cmd = this.comandoLeerTodosPorOrden(conn, idOrden)) {
            ResultSet rs = cmd.executeQuery();

            List<LineaOrdenVenta> modelos = new ArrayList<>();
            while (rs.next()) {
                modelos.add(this.mapearModelo(rs));
            }

            return modelos;
        }
        catch (SQLException e) {
            System.err.println("Error SQL: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }
}
